viewer_clients/web-client/scripts/codecheckerviewer/BugViewer.js

```diff
@@ -462,46 +462,204 @@ function (declare, dom, style, on, query, Memory, Observable, topic, entities,
+      var that = this;
+      var msg = step.msg;
+
+      // The background must be saved BEFORE stack transition.
+      // Calling is in the caller, return is in the called func, not "outside"
+      var highlight = {
+        background : stack.background,
+        iconOverride : undefined
+      };
+
+      function extractFuncName(prefix) {
+        if (msg.startsWith(prefix)) {
+          return msg.replace(prefix, "").replace(/'/g, "");
+        }
+      }
+
+      var func;
+      if (func = extractFuncName("Calling ")) {
+        stack.funcStack.push(func);
+        stack.background = that._highlight_colours[
+          stack.funcStack.length % that._highlight_colours.length];
+
+        highlight.iconOverride = "calling";
+      } else if (msg.startsWith("Entered call from ")) {
+        highlight.iconOverride = "entered_call";
+      } else if (func = extractFuncName("Returning from ")) {
+        if (func === stack.funcStack[stack.funcStack.length - 1]) {
+          stack.funcStack.pop();
+          stack.background = that._highlight_colours[
+            stack.funcStack.length % that._highlight_colours.length];
+
+          highlight.iconOverride = "returning";
+        } else {
+          throw {
+            name : 'StackError',
+            message : "Returned from " + func + " while the last function " +
+                      "was " + stack.funcStack[stack.funcStack.length - 1]
+          };
+        }
+      } else if (msg.startsWith("Assuming the condition")) {
+        highlight.iconOverride = "assume_switch";
+      } else if (msg.startsWith("Assuming")) {
+        highlight.iconOverride = "assume_exclamation";
+      } else if (msg == "Entering loop body") {
+        highlight.iconOverride = "loop_enter";
+      } else if (msg.startsWith("Loop body executed")) {
+        highlight.iconOverride = "loop_execute";
+      } else if (msg == "Looping back to the head of the loop") {
+        highlight.iconOverride = "loop_back";
+      }
+
+      return highlight;
+    },
+
     _addReport : function (report) {
       var that = this;
 
       this.bugStore.put({
         id : report.reportId + '',
         name : 'Line ' + report.lastBugPosition.startLine
-             + ': ' + report.checkerId,
+             + ' &ndash; ' + report.checkerId,
         parent : util.severityFromCodeToString(report.severity),
         report : report,
-        isLeaf : false
+        isLeaf : false,
+        kind : 'bugpath'
       });
 
       this.bugStore.put({
         id : report.reportId + '_0',
-        name : '<b><u>Result</u>: '
-             + entities.encode(report.checkerMsg)
-             + '</b>',
+        name : '<b><u>' + entities.encode(report.checkerMsg) + '</u></b>',
         parent : report.reportId + '',
         report : report,
-        isLeaf : true
+        isLeaf : true,
+        kind : 'msg'
       });
 
       CC_SERVICE.getReportDetails(report.reportId, function (reportDetails) {
```

Comment:
Could you please refactor this lambda to a "private" report detail formatter function? Don't use too big lambdas.